selectrum.el

```diff
@@ -471,15 +418,104 @@ function and BODY opens the minibuffer."
              ,@body)
          (remove-hook 'minibuffer-setup-hook ,hook)))))
 
-;;; Minibuffer state
+;;; Variables
 
-(defvar-local selectrum--last-buffer nil
-  "The buffer that was current before the active session")
+(defvar selectrum-minibuffer-map
+  (let ((map (make-sparse-keymap)))
+    (set-keymap-parent map minibuffer-local-map)
+
+    (define-key map [remap keyboard-quit] #'abort-recursive-edit)
+    ;; This is bound in `minibuffer-local-map' by loading `delsel', so
+    ;; we have to account for it too.
+    (define-key map [remap minibuffer-keyboard-quit]
+      #'abort-recursive-edit)
+    ;; Override both the arrow keys and C-n/C-p.
+    (define-key map [remap previous-line]
+      #'selectrum-previous-candidate)
+    (define-key map [remap next-line]
+      #'selectrum-next-candidate)
+    (define-key map [remap previous-line-or-history-element]
+      #'selectrum-previous-candidate)
+    (define-key map [remap next-line-or-history-element]
+      #'selectrum-next-candidate)
+    (define-key map [remap exit-minibuffer]
+      #'selectrum-select-current-candidate)
+    (define-key map [remap scroll-down-command]
+      #'selectrum-previous-page)
+    (define-key map [remap scroll-up-command]
+      #'selectrum-next-page)
+    ;; Use `minibuffer-beginning-of-buffer' for Emacs >=27 and
+    ;; `beginning-of-buffer' for Emacs <=26.
+    (define-key map [remap minibuffer-beginning-of-buffer]
+      #'selectrum-goto-beginning)
+    (define-key map [remap beginning-of-buffer]
+      #'selectrum-goto-beginning)
+    (define-key map [remap end-of-buffer]
+      #'selectrum-goto-end)
+    (define-key map [remap kill-ring-save]
+      #'selectrum-kill-ring-save)
+    (define-key map [remap previous-matching-history-element]
+      #'selectrum-select-from-history)
+    (define-key map (kbd "C-M-DEL") #'backward-kill-sexp)
+    (define-key map (kbd "C-M-<backspace>") #'backward-kill-sexp)
+    (define-key map (kbd "C-j") #'selectrum-submit-exact-input)
+    (define-key map (kbd "TAB") #'selectrum-insert-current-candidate)
+    (define-key map (kbd "M-q") 'selectrum-cycle-display-style)
+    ;; Return the map.
+    map)
+  "Keymap used by Selectrum in the minibuffer.")
 
 (defvar selectrum--candidates-overlay nil
   "Overlay used to display current candidates.")
 
-(defvar selectrum--dynamic-candidates nil
+(defvar selectrum--count-overlay nil
```

Comment:
The overlay could also be local since it is tied to a specific recursive minibuffer session I suppose?

Reply:
May apply to both overlays? candidates and count?

Reply:
We use `move-overlay` it shouldn't be set locally.

Reply:
You probably looked at the old version? This PR handles them both equally

Reply:
Okay, while I have seen the `move-overlay`, I have only a superficial picture here. But this does not sound like a good design decision. Overlays are tied to a buffer (see the function `overlay-buffer`). When going into minibuffer recursion it is probably better to create a separate overlay for each minibuffer buffer. I am also using buffer-local overlays in Consult and that approach seems to work well.

Reply:
I looked at the current version and checked the move-overlay, where you are moving the overlay to the current buffer. I still don't understand why this is useful. This leads to an unnecessary coupling of recursive minibuffer states. Except for the overlays it seems to be the case that the recursive minibuffer sessions are fully isolated, which I believe to be a good property.

Reply:
Hm this is on purpose, I introduced it because I thought moving the old one is better here because there is really only one overlay to be at a time. I believe I got the idea from icomplete which does the same in my memory.

Reply:
In my mind it makes sense to keep it that way, the overlay is like a canvas each session draws it's state on while it's possible to give each session it's own I see no positives for that.

Reply:
Yes, Icomplete seems to have only a single overlay. I still consider this a wrong design. Maybe there is a technical reason to have only a single one, e.g., if the overlays would somehow interfere with each other. But from my understanding the overlays are tied to the buffer without interference. Otherwise I would have to rework some things in Consult too :sweat_smile: For this reason I am also interested in how this turns out here. So if there is no technical reason to have buffer local overlays, I would go with that since the isolation will help with overall maintenance and it is just a cleaner design. The only technical reason which may be there is that having a single overlay saves resources, but this argument is pretty much of no relevance. But maybe there is another argument?

Reply:
I have sometimes observed an issue with recursive minibuffer sessions where the old minibuffer session is not updated directly when leaving the old minibuffer, which seemed like a bug to me. Maybe this is related to the fact that you are only using a single overlay? If the old overlay in the old minibuffer session would still be there wouldn't be such an updating issue. But this is a bit of a speculation. I am not sure if this is really due the overlay.

Reply:
We have to agree to disagree here, if you wouldn't use an overlay you would erase things and redraw because you have a single canvas not multiple ones. I prefer to have it that way but I also think it is perfectly fine to do it like you suggest.

Reply:
> I have sometimes observed an issue with recursive minibuffer sessions where the old minibuffer session is not updated directly when leaving the old minibuffer, which seemed like a bug to me. Maybe this is related to the fact that you are only using a single overlay? 

I did not oberver that but that would be a compelling argument to change it indeed, can you reproduce it?

Reply:
> if you wouldn't use an overlay you would erase things and redraw because you have a single canvas not multiple ones.

I don't understand what you mean by that. If you use one overlay per recursive session you wouldn't have to erase or redraw anything?

>  I prefer to have it that way but I also think it is perfectly fine to do it like you suggest.

For which reason, if I may ask? I gave arguments why I consider it better to use the isolation and I am also using the isolation approach in Consult. It would be interesting to see why you prefer it that way. But sure, ultimately it is your project and you decide.

Reply:
I gave the reason multiple times but it seems you don't accept it as a reason: A single canvas. You can't draw the different overlays at the same time so there is no reason to pretend you can if that make sense, there is no need to isolate anything here in my opinion.

Reply:
> I don't understand what you mean by that. If you use one overlay per recursive session you wouldn't have to erase or redraw anything?

I meant it you wouldn't use an overlay at all an would like to draw things in a buffer and then update that buffer according to change state you would erase it and reinsert.

Reply:
Ah I see, I didn't get this argument, since it goes against the logic of how Emacs overlays and buffers work. So yes, you could say that I don't accept this as a reason ;) I think it is just more consistent to go with a single overlay per buffer, Emacs also uses separate buffers for recursive minibuffers despite there being only a single canvas.

Regarding the issue I described above, which *could* be due to the overlay reuse, the mere possibility of introducing bugs by not fully isolating the sessions is reason enough for me to go with the full isolation. I will try to reproduce the issue then I can make a more well-informed argument.

Reply:
BTW I always have weird delays in these review comments for whatever reason.

Reply:
> BTW I always have weird delays in these review comments for whatever reason.

You mean github not updating? Yes for me the comments also often don't update. This can lead to confusing out-of-order discussions.

Reply:
I also like to avoid possibilities for bugs of course but I'm not sure what you observe is likely to have something to with this the overlay is still attached only to a single buffer at a time.

Reply:
Okay, so the problem I am observing can be reproduced as follows:

emacs -Q, Emacs 27.1

~~~ elisp
(package-initialize)
(require 'selectrum)
(selectrum-mode)
(setq enable-recursive-minibuffers t)
~~~

Press `M-x C-x o M-x C-g` and wait. Then the Selectrum minibuffer is not updated.

Reply:
Nice one, I can reproduce it and stand corrected ;) I will change to overlays per session then.

Reply:
Thanks! :sweat_smile:

Reply:
Note that you could still go with the single overlay, it is just *much harder*. I think everywhere where you have this shared state, there is a small bug lurking. There is a reason why I am pushy about #322 :laughing:

Reply:
No thank you! :)

Reply:
Icomplete seems to also have this updating bug btw. Probably for the same reason.

Reply:
One could add to the README comparison list - Selectrum handles recursive sessions properly ;)

Reply:
Yes, seems likely now. I don't often use recursive sessions that is why I never observed it I guess. So this is a clear win for your isolation argument.

Reply:
With recursive sessions I feel it would be much nicer if Emacs would behave like `embark-become` and simply close the other session. I'm always annoyed when doing something in a recursive session (like opening help buffer from drescribe-function) only to see it closed when exiting the session...

Reply:
Well, but then it wouldn't be recursive ;) But I am also not entirely convinced by recursive editing. For example with consult-line one can jump out of the minibuffer, do some editing and jump back. If that's useful or not ¯\_(ツ)_/¯